crates/core_simd/src/swizzle.rs

```diff
@@ -0,0 +1,374 @@
+use crate::simd::intrinsics;
+use crate::{LaneCount, Simd, SimdElement, SupportedLaneCount};
+
+/// Rearrange vector elements.
+///
+/// A new vector is constructed by specifying the the lanes of the source vector or vectors to use.
+///
+/// When swizzling one vector, the indices of the result vector are indicated by a `const` array
+/// of `usize`, like [`Swizzle`].
+/// When swizzling two vectors, the indices are indicated by a `const` array of [`Which`], like
+/// [`Swizzle2`].
+///
+/// # Examples
+/// ## One source vector
+/// ```
+/// # #![feature(portable_simd)]
+/// # use core_simd::{Simd, simd_swizzle};
+/// let v = Simd::<f32, 4>::from_array([0., 1., 2., 3.]);
+///
+/// // Keeping the same size
+/// let r = simd_swizzle!(v, [3, 0, 1, 2]);
+/// assert_eq!(r.to_array(), [3., 0., 1., 2.]);
+///
+/// // Changing the number of lanes
+/// let r = simd_swizzle!(v, [3, 1]);
+/// assert_eq!(r.to_array(), [3., 1.]);
+/// ```
+///
+/// ## Two source vectors
+/// ```
+/// # #![feature(portable_simd)]
+/// # use core_simd::{Simd, simd_swizzle, Which};
+/// use Which::*;
+/// let a = Simd::<f32, 4>::from_array([0., 1., 2., 3.]);
+/// let b = Simd::<f32, 4>::from_array([4., 5., 6., 7.]);
+///
+/// // Keeping the same size
+/// let r = simd_swizzle!(a, b, [First(0), First(1), Second(2), Second(3)]);
+/// assert_eq!(r.to_array(), [0., 1., 6., 7.]);
+///
+/// // Changing the number of lanes
+/// let r = simd_swizzle!(a, b, [First(0), Second(0)]);
+/// assert_eq!(r.to_array(), [0., 4.]);
+/// ```
+#[macro_export]
+macro_rules! simd_swizzle {
+    {
+        $vector:expr, $index:expr $(,)?
+    } => {
+        {
+            use $crate::simd::Swizzle;
+            struct SwizzleImpl;
+            impl<const LANES: usize> Swizzle<LANES, {$index.len()}> for SwizzleImpl {
+                const INDEX: [usize; {$index.len()}] = $index;
+            }
+            SwizzleImpl::swizzle($vector)
+        }
+    };
+    {
+        $first:expr, $second:expr, $index:expr $(,)?
+    } => {
+        {
+            use $crate::simd::{Which, Swizzle2};
+            struct SwizzleImpl;
+            impl<const LANES: usize> Swizzle2<LANES, {$index.len()}> for SwizzleImpl {
+                const INDEX: [Which; {$index.len()}] = $index;
+            }
+            SwizzleImpl::swizzle2($first, $second)
+        }
+    }
+}
+
+/// An index into one of two vectors.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum Which {
+    /// Indexes the first vector.
+    First(usize),
+    /// Indexes the second vector.
+    Second(usize),
+}
```

Comment:
It is traditional to name this type "Either", and its variants are named "Left" and "Right".
I am not a particularly devout adherent of tradition, however, so I am fine with whatever, at the moment.

Reply:
Agreed, now that I look at it, it seems like this really should just be an `Either`.  Which isn't really the place of `core::simd` to add, but it's by far a superior API in this case.

Reply:
Normally in Rust we use Result for what is called Either, because in languages that have Either, they implicitly use it as a Result, but sometimes it really is Either. So I think it is harmless if we have `core::simd::Either`: it's specifically for this library, after all.

Reply:
I think maybe it's not quite correct to call it `Either`, since that would imply that it should be `Either<usize>`, I think.

Reply:
That would be `Maybe<usize>`, no?
I am not of a strong opinion however, here.